core/src/main/java/tc/oc/pgm/shops/ShopModule.java

```diff
@@ -0,0 +1,217 @@
+        }
+        shopKeeperLocations.stream()
+            .map(
+                data ->
+                    new ShopKeeper(
+                        data.getName(),
+                        data.getLocation(),
+                        data.getYaw(),
+                        data.getPitch(),
+                        data.getType(),
+                        shopInstance))
+            .forEach(keeper -> module.shopKeepers.add(keeper));
+      }
+
+      return !module.shops.isEmpty() ? module : null;
+    }
+  }
+
+  private static SetMultimap<String, KeeperData> parseShopKeepers(Element root, Logger logger)
+      throws InvalidXMLException {
+    SetMultimap<String, KeeperData> shopkeepers = HashMultimap.create();
+
+    for (Element shopkeeper : XMLUtils.flattenElements(root, "shopkeepers")) {
+      Attribute locationAttr = XMLUtils.getRequiredAttribute(shopkeeper, "location");
+      Attribute yawAttr = XMLUtils.getRequiredAttribute(shopkeeper, "yaw");
+      Attribute pitchAttr = XMLUtils.getRequiredAttribute(shopkeeper, "pitch");
+      Attribute typeAttr = XMLUtils.getRequiredAttribute(shopkeeper, "type");
+
+      String name = XMLUtils.getNullableAttribute(shopkeeper, "name");
+      Vector location = XMLUtils.parseVector(locationAttr);
+      Float yaw = XMLUtils.parseNumber(yawAttr, Float.class, 0f);
+      Float pitch = XMLUtils.parseNumber(pitchAttr, Float.class, 0f);
+      Class<? extends Entity> mob =
+          XMLUtils.parseEntityTypeAttribute(shopkeeper, "mob", Villager.class);
+
+      shopkeepers.put(typeAttr.getValue(), new KeeperData(name, location, yaw, pitch, mob));
+    }
+
+    return shopkeepers;
+  }
+
+  private static List<Icon> parseIcons(Element category, KitParser kits, Logger logger)
+      throws InvalidXMLException {
+    List<Icon> icons = Lists.newArrayList();
+    for (Element icon : XMLUtils.getChildren(category, "item")) {
+      icons.add(parseIcon(icon, kits, logger));
+    }
+
+    if (icons.size() > Category.MAX_ICONS) {
+      throw new InvalidXMLException(
+          "Categories may only contain " + Category.MAX_ICONS + " icons", category);
+    }
+
+    if (icons.isEmpty()) {
+      throw new InvalidXMLException("At least one <item> is required per category", category);
+    }
+
+    return icons;
+  }
+
+  private static Icon parseIcon(Element icon, KitParser kits, Logger logger)
+      throws InvalidXMLException {
+    Material currency = parseMaterial(XMLUtils.getRequiredAttribute(icon, "currency"), icon);
+    Integer price =
+        XMLUtils.parseNumber(XMLUtils.getRequiredAttribute(icon, "price"), Integer.class);
+    ItemStack item = kits.parseItem(icon, false);
+
+    Kit kit = null;
+    Attribute kitAttr = icon.getAttribute("kit");
+    if (kitAttr != null) {
+      kit = kits.parseKitProperty(icon, "kit", null);
+    } else {
+      kit = new ItemKit(Maps.newHashMap(), Lists.newArrayList(item));
```

Comment:
^

Reply:
Could you elaborate a bit more on this when you have a chance? Still new to Actions, would this be simply parsing with the `ActionParser` or something else?

Reply:
Instead of parsing as a kit, parse as an action, the action parser probably either has, or should have a similar to "parseKitProperty"

Regarding making your own to just give items, yeah, you just create a class which implements Action, and define the method to "trigger"

Reply:
Ah fantastic! Thanks for explaining, makes sense 👍 I'll see if I can get that done soon

Reply:
@Pablete1234 there's `ActionParser#parseKitTrigger`, though believe it won't parse the same way as `parseKitProperty` does. Would it be fine to define a method like `parseKitPropertyTrigger` inside `ActionParser` or is there a different approach you'd suggest? 

As for the dummy action, would something like this work?
```java
public class GiveItemAction extends AbstractAction<MatchPlayer> {

  private final ItemStack item;

  public GiveItemAction(ItemStack item) {
    super(MatchPlayer.class);
    this.item = item;
  }

  @Override
  public void trigger(MatchPlayer player) {
    HashMap<Integer, ItemStack> overflow = player.getPlayer().getInventory().addItem(item);

    // Drop item at feet
    if (!overflow.isEmpty()) {
       // will be translatable in production
      player.sendWarning(text("Purchase could not fit in your inventory, so it's been dropped at your feet"));
      World world = player.getWorld();
      overflow.values().forEach(stack -> world.dropItemNaturally(player.getLocation(), stack));
    }
  }
}
```

Reply:
`ActionParser#parseReference` is what you're looking for, it will return an Action of the bound you ask for:

`Action<MatchPlayer> action = actionParser.parseReference(Node.forAttribute("action"), null, MatchPlayer.class)`
Keep in mind, this action **can** be a kit, because kits implement `Action<MatchPlayer>` so it all ties up together nicely.

Regarding that sample action, yeah it looks good, i'd actually make the "drop-overflow" an optional parameter in the constructor, and expose it as a newly available action:
```xml
<give drop-overflow="false">
    <item ...>
    <item ...>
</give-item>
```
This allows for giving items to a player but without following a kit's logic to giving items (where they won't give more if you already have the item)
It may be out of the scope of this but it'd be nice to have